cmd/remote_val_harness/internal/test_harness.go

```diff
@@ -0,0 +1,402 @@
+		sc:               sc,
+		fpv:              fpv,
+		chainID:          st.ChainID,
+		acceptRetries:    cfg.AcceptRetries,
+		logger:           logger,
+		exitWhenComplete: cfg.ExitWhenComplete,
+		exitCode:         0,
+	}, nil
+}
+
+// Run will execute the tests associated with this test harness. The intention
+// here is to call this from one's `main` function, as the way it succeeds or
+// fails at present is to call os.Exit() with an exit code related to the error
+// that caused the tests to fail, or exit code 0 on success.
+func (th *TestHarness) Run() {
+	th.logger.Info("Starting test harness")
+	donec := make(chan struct{})
+	go func() {
+		defer close(donec)
+		accepted := false
+		var startErr error
+		for acceptRetries := th.acceptRetries; acceptRetries > 0; acceptRetries-- {
+			th.logger.Info("Attempting to accept incoming connection", "acceptRetries", acceptRetries)
+			if err := th.sc.Start(); err != nil {
+				// if it wasn't a timeout error
+				if _, ok := err.(timeoutError); !ok {
+					th.logger.Error("Failed to start listener", "err", err)
+					th.Shutdown(newTestHarnessError(ErrFailedToStartListener, err, ""))
+					// we need the return statements in case this is being run
+					// from a unit test - otherwise this function will just die
+					// when os.Exit is called
+					return
+				}
+				startErr = err
+			} else {
+				accepted = true
+				break
+			}
+		}
+		if !accepted {
+			th.logger.Error("Maximum accept retries reached", "acceptRetries", th.acceptRetries)
+			th.Shutdown(newTestHarnessError(ErrMaxAcceptRetriesReached, startErr, ""))
+			return
+		}
+
+		// Run the tests
+		if err := th.TestPublicKey(); err != nil {
+			th.Shutdown(err)
+			return
+		}
+		if err := th.TestSignProposal(); err != nil {
+			th.Shutdown(err)
+			return
+		}
+		if err := th.TestSignVote(); err != nil {
+			th.Shutdown(err)
+			return
+		}
+		th.logger.Info("SUCCESS! All tests passed.")
+		th.Shutdown(nil)
+	}()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	go func() {
+		for sig := range c {
+			th.logger.Info("Caught interrupt, terminating...", "sig", sig)
+			th.Shutdown(newTestHarnessError(ErrInterrupted, nil, ""))
+		}
+	}()
+
+	// Run until complete
+	<-donec
+}
+
+// TestPublicKey just validates that we can (1) fetch the public key from the
+// remote signer, and (2) it matches the public key we've configured for our
+// local Tendermint version.
+func (th *TestHarness) TestPublicKey() error {
+	th.logger.Info("TEST: Public key of remote signer")
+	th.logger.Info("Local", "pubKey", th.fpv.GetPubKey())
+	th.logger.Info("Remote", "pubKey", th.sc.GetPubKey())
+	if th.fpv.GetPubKey() != th.sc.GetPubKey() {
+		th.logger.Error("FAILED: Local and remote public keys do not match")
+		return newTestHarnessError(ErrTestPublicKeyFailed, nil, "")
+	}
+	return nil
+}
+
+// TestSignProposal makes sure the remote signer can successfully sign
+// proposals.
+func (th *TestHarness) TestSignProposal() error {
```

Comment:
We later should add failing tests too. If the proposal is not valid for instance. Not in this PR though.